File: PID/backend/microservices/Evenement/src/main/java/org/example/evenement/service/InscriptionEvenementService.java
package org.example.evenement.service;

import org.example.evenement.dto.InscriptionResultDto;
import org.example.evenement.dto.StatutInscriptionDto;
import org.example.evenement.entity.InscriptionEvenement;

import java.util.List;
import java.util.Optional;

public interface InscriptionEvenementService {
    
    // Créer une nouvelle inscription (ou ajouter en liste d'attente si complet)
    InscriptionResultDto createInscription(InscriptionEvenement inscription);
    
    // Récupérer toutes les inscriptions
    List<InscriptionEvenement> getAllInscriptions();
    
    // Récupérer une inscription par son ID
    Optional<InscriptionEvenement> getInscriptionById(Long id);
    
    // Mettre à jour une inscription
    InscriptionEvenement updateInscription(Long id, InscriptionEvenement inscription);
    
    // Supprimer une inscription
    void deleteInscription(Long id);
    
    // Récupérer les inscriptions d'un étudiant
    List<InscriptionEvenement> getInscriptionsByEtudiant(Long idEtudiant);
    
    // Récupérer les inscriptions d'un événement
    List<InscriptionEvenement> getInscriptionsByEvenement(Long evenementId);
    
    // Vérifier si un étudiant est inscrit à un événement
    boolean isEtudiantInscrit(Long idEtudiant, Long evenementId);

    // Vérifier si un étudiant est en liste d'attente pour un événement
    boolean isEtudiantEnListeAttente(Long idEtudiant, Long evenementId);

    // Obtenir la position en liste d'attente (0 si non en liste)
    int getPositionListeAttente(Long idEtudiant, Long evenementId);

    // Obtenir le statut complet (inscrit, liste attente, position)
    StatutInscriptionDto getStatutInscription(Long idEtudiant, Long evenementId);
    
    // Compter les inscriptions pour un événement
    long countInscriptionsByEvenement(Long evenementId);
    
    // Désinscrire un étudiant d'un événement
    void desinscrireEtudiant(Long idEtudiant, Long evenementId);
}

File: PID/frontend/src/app/services/inscription-evenement.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { InscriptionEvenement, InscriptionResultDto, StatutInscriptionDto } from '../models/evenement.model';
import { API_URL } from '../api.config';

@Injectable({
  providedIn: 'root'
})
export class InscriptionEvenementService {
  private apiUrl = `${API_URL}/inscriptions`;

  constructor(private http: HttpClient) { }

  // CREATE - Créer une nouvelle inscription (ou liste d'attente si complet)
  createInscription(inscription: InscriptionEvenement): Observable<InscriptionResultDto> {
    return this.http.post<InscriptionResultDto>(this.apiUrl, inscription);
  }

  // READ - Récupérer toutes les inscriptions
  getAllInscriptions(): Observable<InscriptionEvenement[]> {
    return this.http.get<InscriptionEvenement[]>(this.apiUrl);
  }

  // READ - Récupérer une inscription par son ID
  getInscriptionById(id: number): Observable<InscriptionEvenement> {
    return this.http.get<InscriptionEvenement>(`${this.apiUrl}/${id}`);
  }

  // UPDATE - Mettre à jour une inscription
  updateInscription(id: number, inscription: InscriptionEvenement): Observable<InscriptionEvenement> {
    return this.http.put<InscriptionEvenement>(`${this.apiUrl}/${id}`, inscription);
  }

  // DELETE - Supprimer une inscription
  deleteInscription(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }

  // Récupérer les inscriptions d'un étudiant
  getInscriptionsByEtudiant(idEtudiant: number): Observable<InscriptionEvenement[]> {
    return this.http.get<InscriptionEvenement[]>(`${this.apiUrl}/etudiant/${idEtudiant}`);
  }

  // Récupérer les inscriptions d'un événement
  getInscriptionsByEvenement(evenementId: number): Observable<InscriptionEvenement[]> {
    return this.http.get<InscriptionEvenement[]>(`${this.apiUrl}/evenement/${evenementId}`);
  }

  // Vérifier si un étudiant est inscrit à un événement
  isEtudiantInscrit(idEtudiant: number, evenementId: number): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/verifier/${idEtudiant}/${evenementId}`);
  }

  getStatutInscription(idEtudiant: number, evenementId: number): Observable<StatutInscriptionDto> {
    return this.http.get<StatutInscriptionDto>(`${this.apiUrl}/statut/${idEtudiant}/${evenementId}`);
  }

  // Compter les inscriptions pour un événement
  countInscriptionsByEvenement(evenementId: number): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/count/evenement/${evenementId}`);
  }

  // Désinscrire un étudiant d'un événement
  desinscrireEtudiant(idEtudiant: number, evenementId: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/desinscrire/${idEtudiant}/${evenementId}`);
  }
}

